src/lib.rs

```diff
@@ -1,27 +1,56 @@
-use std::collections::HashMap;
-
 use borsh::schema::{BorshSchemaContainer, Declaration, Definition, Fields, VariantName};
 use schemars::schema::{RootSchema, Schema};
-use serde::{Deserialize, Serialize};
+use semver::Version;
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
 
 #[doc(hidden)]
 #[cfg(feature = "__chunked-entries")]
 #[path = "private.rs"]
 pub mod __private;
 
+// Keep in sync with SCHEMA_VERSION below.
+const SCHEMA_SEMVER: Version = Version {
+    major: 0,
+    minor: 1,
+    patch: 0,
+    pre: semver::Prerelease::EMPTY,
+    build: semver::BuildMetadata::EMPTY,
+};
+
 /// Current version of the ABI schema format.
-pub const SCHEMA_VERSION: &str = env!("CARGO_PKG_VERSION");
+pub const SCHEMA_VERSION: &str = "0.1.0";
 
 /// Contract ABI.
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
 pub struct AbiRoot {
```

Comment:
What do we think about adding this – `#[serde(deny_unknown_fields)]`?

And then the order of errors would be:

1. schema version check
2. ensure no unknown fields
3. ensure known fields are all present

(doesn't have to happen in this PR)

Reply:
Yeah makes sense to me. I would have to add this to all models, not just the top-level one, right?

Reply:
Well, all models except `AbiMetadata`. You seem to have a `#[serde(flatten)] HashMap<String, String>` for catching arbitrary metadata inputs.